modules/local_contacts/modules.php

```diff
@@ -68,6 +68,62 @@ public function process() {
             $this->session->record_unsaved('Contact Added');
             Hm_Msgs::add('Contact Added');
         }
+
+        list($success, $form) = $this->process_form(array('contact_source', 'add_contact'));
+        if ($success && $form['contact_source'] == 'csv') {
+            $file = $this->request->files['contact_csv'];
+            $csv = fopen($file['tmp_name'], 'r');
+            if ($csv) {
+                $contacts = $this->get('contact_store');
+                $message = '';
+                $header = fgetcsv($csv);
+                $expectedHeader = array('display_name', 'email_address', 'phone_number');
+
+                if ($header !== $expectedHeader) {
+                    fclose($csv);
+                    Hm_Msgs::add('ERRInvalid CSV file, please use a valid header: '.implode(', ', $expectedHeader));
+                    return;
+                }
+
+                $contact_list = $this->user_config->get('contacts', array());
+                $contact_list = array_map(function($v) { $v['type'] = 'local'; return $v; }, $contact_list);
+
+                while (($data = fgetcsv($csv)) !== FALSE) {
+                    $email = $data[1];
+                    if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
+                        Hm_Msgs::add('ERRInvalid email address: '.$email.'\n');
+                        continue;
+                    }
+
+                    $details = array('source' => 'local', 'display_name' => $data[0], 'email_address' => $email);
+                    if (array_key_exists(2, $data) && $data[2]) {
+                        $details['phone_number'] = $data[2];
+                    }
+
+                    $contactUpdated = false;
+                    
+                    foreach ($contact_list as $key => $contact) {
+                        if ($contact['email_address'] == $email) {
+                            $contacts->update_contact($key, $details);
+                            $message .= 'Contact Updated: '.$email.'\n';
+                            $contactUpdated = true;
+                            continue 2;
+                        }
+                    }
+
+                    if (!$contactUpdated) {
+                        $contacts->add_contact($details);
+                        $message .= 'Contact Added: '.$email.'\n';
+                    }
+                }
+                fclose($csv);
+                $this->user_config->set('contacts', $contacts->export());
+                $this->session->record_unsaved('Contact Added');
+                if ($message) {
+                    Hm_Msgs::add($message);
```

Comment:
Have you tested importing 100 or 1000 contacts? I think this message might not fit in the hm messages elegantly in that case. Maybe it is worth adding a separate output module to display the results of the import operation - one line per the input file with a success mark/message (e.g. created or updated) or error message if the line didn't import.